tests/shared/src/test/scala/munit/AsyncFixtureSuite.scala

```diff
@@ -0,0 +1,113 @@
+package munit
+
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class AsyncFixtureSetupTestTeardownOrderSuite extends FunSuite {
+  val latch = Promise[Unit]
+  var completedFromTest = Option.empty[Boolean]
+  var completedFromTeardown = Option.empty[Boolean]
+
+  val latchOnTeardown = FunFixture.async[String](
+    setup = { test => Future.successful(test.name) },
+    teardown = { name =>
+      implicit val ec = munitExecutionContext
+      Future {
+        completedFromTeardown = Some(latch.trySuccess(()));
+      }
+    }
+  )
+
+  override def afterAll(): Unit = {
+    assertEquals(completedFromTest, Some(true))
+    assertEquals(completedFromTeardown, Some(false))
+  }
+
+  latchOnTeardown.test("test is ran before the teardown") { _ =>
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future {
+      // Ideally we would simulate some work here, for example by using Thread.sleep(50),
+      // which would increase the certainty that this test passes by design and not by lucky
+      // scheduling. However Thread.sleep is not scala-native compatible.
```

Comment:
As it stands right now, this test-case will always pass, because teardown execution is flat-mapped over the test body. But if this happens-before relationship is removed inadvertently, this testcase can still pass by scheduler luck. Moving this test-case to JVM-only sounds good.